homeassistant/components/discovergy/sensor.py

```diff
@@ -0,0 +1,274 @@
+            ) from err
+        except HTTPError as err:
+            raise UpdateFailed(f"Error communicating with API: {err}") from err
+        except Exception as err:  # pylint: disable=broad-except
+            raise UpdateFailed(
+                f"Unexpected error while communicating with API: {err}"
+            ) from err
+
+    coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        name="sensor",
+        update_method=async_update_data,
+        update_interval=update_interval,
+    )
+    return coordinator
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up the Discovergy sensors."""
+    data: DiscovergyData = hass.data[DOMAIN][entry.entry_id]
+    discovergy_instance: Discovergy = data.api_client
+    meters: list[Meter] = data.meters  # always returns a list
+
+    min_time_between_updates = timedelta(
+        seconds=entry.options.get(CONF_TIME_BETWEEN_UPDATE, DEFAULT_TIME_BETWEEN_UPDATE)
+    )
+
+    entities = []
+    for meter in meters:
+        # Get coordinator for meter, set config entry and fetch initial data
+        # so we have data when entities are added
+        coordinator = get_coordinator_for_meter(
+            hass, meter, discovergy_instance, min_time_between_updates
+        )
+        coordinator.config_entry = entry
+        await coordinator.async_config_entry_first_refresh()
+
+        # add coordinator to data for diagnostics
+        data.coordinators[meter.get_meter_id()] = coordinator
+
+        sensors = None
+        if meter.measurement_type == "ELECTRICITY":
+            sensors = ELECTRICITY_SENSORS
+        elif meter.measurement_type == "GAS":
+            sensors = GAS_SENSORS
+
+        if sensors is not None:
+            for description in sensors:
+                keys = [description.key] + description.alternative_keys
+
+                # check if this meter has this data, then add this sensor
+                for key in keys:
+                    if key in coordinator.data.values:
+                        entities.append(
+                            DiscovergySensor(key, description, meter, coordinator)
+                        )
+
+    async_add_entities(entities, False)
+
+
+class DiscovergySensor(CoordinatorEntity, SensorEntity):
+    """Represents a discovergy smart meter sensor."""
+
+    entity_description: DiscovergySensorEntityDescription
+    data_key: str
+    _attr_has_entity_name = True
+
+    def __init__(
+        self,
+        data_key: str,
+        description: DiscovergySensorEntityDescription,
+        meter: Meter,
+        coordinator: DataUpdateCoordinator,
+    ) -> None:
+        """Initialize the sensor."""
+        super().__init__(coordinator)
+
+        self.data_key = data_key
+
+        self.entity_description = description
+        self._attr_unique_id = f"{meter.full_serial_number}-{description.key}"
```

Comment:
This won't be unique as there will potentially be more than one entity per description via the alternative_keys. Probably use the `data_key` as suffix instead.